[samuelkwik] iP

samuelkwik opened the pull request:
### Viscount 
#### _the okayish chatbot_

> "Only mediocrity can be trusted to be always at its best." – Sir Max Beerbohm

It's not yet a Duke, meet Viscount! For when you already have a calendar app in your pocket but want to make life a little more cumbersome.

Viscount, a blast from the past, is:
- entirely text-based
- ~not really~ intuitive
- okay

All you need to do to get started down a nostalgic trip is,

1. download it [here](https://github.com/samuelkwik/ip)!
2. figure out how to run compile and run it.
3. add your tasks.
4. let it manage it for you the best that it can!

And best of all, it is perpetually **FREE**! ~On the condition that no sane company buys the rights to this and decides that a subscription model is best for users moving forward.~

With a cutting-edge feature list circa 1950,

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Pretending to be your imaginary friend (coming never)

***
If you are a Java programmer (why?), you can use it to practice Java too. Here's the `main` method: 

``` java
public static void main(String[] args) {
    Viscount viscount = new Viscount();
    viscount.run();
}
```

What are you waiting for? Use Viscount now! :sparkles:

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

samuelkwik committed:
Rename "Duke" to "Viscount"

samuelkwik committed:
Add greeting and exit

samuelkwik committed:
Add support for "Echo"

samuelkwik committed:
Changed "echo" functionality into list add/view

samuelkwik committed:
Renamed Viscount.java to viscount.java and moved it to it's own package

samuelkwik committed:
Created "TaskList.java" to abstract task list away from main

samuelkwik committed:
Removed all instances of Duke and corrected capitalisation

samuelkwik committed:
Add Task class to encapsulate functions such as toggling task doneness and formatting

samuelkwik committed:
Add support to toggle task between complete and incomplete

samuelkwik committed:
Viscount.java: Rename parseInput method to getCommand method

samuelkwik committed:
Move Command enum from Viscount.java to Command.java

samuelkwik committed:
Move Command enum to Command.java

samuelkwik committed:
Revert "Move Command enum to Command.java"

This reverts commit 2968637054be3470fc09dcefbe5242ba5bdf6de5.

samuelkwik committed:
Move enum to own file

samuelkwik committed:
Added ParsedCommand class to encapsulate commands and required argument strings

samuelkwik committed:
Add support for to dos

samuelkwik committed:
Add custom exception for chatbot exceptions

samuelkwik committed:
Add support for deadlines

samuelkwik committed:
Added support for Event

samuelkwik committed:
Fix bugs related to Event: event wrongly captured as deadline

samuelkwik committed:
Fix trailing spaces being saved in task fields

samuelkwik committed:
Standardise error for all tasks to include usage tip

samuelkwik committed:
Modify TextUI testing scripts and EXPECTED.txt

samuelkwik committed:
Fix formatting

samuelkwik committed:
Add support for deleting tasks and updated UI-testing to reflect new function

samuelkwik committed:
Add support for automatic file storage and retrieval upon update of taskList

samuelkwik committed:
Merge branch 'branch-Level-7'

samuelkwik committed:
Add support for dates for tasks containing date fields

samuelkwik committed:
Merge branch 'branch-Level-8'

samuelkwik committed:
Refactor Viscount into separate classes handling UI, Storage and Parsing

samuelkwik committed:
Add parent abstract class Command and child commands to replace enum Command class

samuelkwik committed:
Remove redundant constructors and parameters from Parser

samuelkwik committed:
Add viscount.command and viscount.task packages

samuelkwik committed:
Merge branch 'add-gradle-support'

samuelkwik committed:
Add gradle support

samuelkwik committed:
Add jUnit test Parser: handleToggle() and ToDo.java

samuelkwik committed:
Fix deletion bug refusing to delete last record

samuelkwik committed:
Format all files to match coding standard

samuelkwik committed:
Add support for find functionality

samuelkwik committed:
Merge branch 'branch-Level-9'

samuelkwik committed:
Fix missing message header for find results

samuelkwik committed:
Add javadoc for ListCommand.java

samuelkwik committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/viscount/command/ListCommand.java

samuelkwik committed:
Fix formatting for FindCommand

samuelkwik committed:
Merge branch 'master' into branch-A-JavaDoc

samuelkwik committed:
Add JavaDoc for all remaining classes

samuelkwik committed:
Merge branch 'branch-A-JavaDoc'

leewaikin19 commented on src/main/java/viscount/task/Event.java:
```diff
@@ -0,0 +1,86 @@
+     * @return A string representing the Event task, including its type, completion status,
+     * description, and start/end dates.
+     */
+    @Override
```
Perhaps it would be useful and easier to use the `@inheritDoc` tag?

samuelkwik replied:
Thank you for this! I honestly did not know that the `@inheritDoc` tag exists... you can probably imagine the pain it was to ensure consistency amongst all the child classes! Thanks again! 😄

leewaikin19 commented on src/main/java/viscount/command/DeleteCommand.java:
```diff
@@ -0,0 +1,51 @@
+     * @param indexStr The index of the task to be deleted, provided as a string.
+     */
+    public DeleteCommand(String indexStr) {
+        this.indexStr = indexStr + "";
```
Is there a need to add `+ "";`?

samuelkwik replied:
Thank you for pointing that out! It is definitely not supposed to be there.

leewaikin19 commented on src/main/java/viscount/command/Parser.java:
```diff
@@ -0,0 +1,103 @@
+            case "todo" -> handleTodo(inputString);
+            case "deadline" -> handleDeadline(inputString);
+            case "event" -> handleEvent(inputString);
+            case "list" -> new ListCommand();
```
A suggestion would be to check for parameters after the `list` and `bye` command, to ensure the user does not misunderstand or misuse the command

samuelkwik replied:
Good suggestion! Thank you!

leewaikin19 commented on src/main/java/viscount/command/UnknownCommand.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Represents a command that handles invalid or unrecognized user input.
+ */
+public class UnknownCommand extends Command {
```
Great use of a separate `UnknownCommand` class to handle invalid cases!

samuelkwik replied:
😸 Thank you!

samuelkwik committed:
Fix formatting for Storage class

samuelkwik committed:
Add early support for JavaFX gui

samuelkwik committed:
Merge branch 'branch-Level-10'

samuelkwik committed:
Added assertions to validate internal states in AddCommand and TaskList

Assertions were introduced to ensure correctness of internal states and to safeguard against potential logical errors during task addition and management.

The added assertions improve code reliability, making it easier to identify potential issues during development or future maintenance.

samuelkwik committed:
Refactor AddCommand and Parser for improved code quality

samuelkwik committed:
Merge pull request #1 from samuelkwik/branch-A-Assertions

Added assertions to validate internal states in AddCommand and TaskList

samuelkwik committed:
Merge branch 'master' into branch-A-CodeQuality

samuelkwik committed:
Merge branch 'branch-A-CodeQuality'

samuelkwik committed:
Add support for undoing add, delete, and toggle commands

Undo command undoes changes made to the taskList through add, delete or toggle commands, with maximum history being one command.

Undo command gives user a safety-net to revert any accidental changes, giving them peace of mind.

samuelkwik committed:
Merge branch 'branch-C-Undo'

samuelkwik committed:
Change user and chatbot display pictures

Change user and chatbot display avatars to add personality to the chatbot.

samuelkwik committed:
Fix GUI elements

Fixed user input field anchoring to have static height when resizing chatbot window.

samuelkwik committed:
Add Ui.png

samuelkwik committed:
Add user guide

samuelkwik committed:
Update README.md

samuelkwik committed:
Update JavaDoc for commands

samuelkwik committed:
Merge branch 'master' of https://github.com/samuelkwik/ip

samuelkwik committed:
Fix unused imports

samuelkwik committed:
Fix typo in README.md

samuelkwik committed:
Fix typo in README.md

samuelkwik committed:
Merge remote-tracking branch 'origin/master'

samuelkwik committed:
Fix typo in README

samuelkwik committed:
Update README.md with AI acknowledgements

samuelkwik committed:
Fix README.md table

samuelkwik committed:
Update README.md
